### PR TITLE
separate rehash query for each partition

### PR DESCRIPTION
**Pull request by katherinebernstein**
The rehashing for BTREE datetimes finished (yay!) so the only ones left to migrate are the not indexed ones

### CODE REVIEW

**Review by you-leee — APPROVED** [review]

**Review by makosblade — APPROVED** [review]

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 1]
```diff
@@ -23,127 +19,201 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
...
 
-        return "INSERT INTO ${DATA.name} ( $insertCols ) " +
-                "SELECT DISTINCT ON ($keyCols) $selectDistinctCols, $selectMetadataCols " +
+        return "INSERT INTO $TEMP_TABLE_NAME " +
+                "SELECT $unchangedCols, $newHashComputation AS ${HASH.name}, ARRAY[${HASH.name}] AS ${OLD_HASHES_COL.name} " +
                 "FROM ${DATA.name} " +
                 "WHERE ${PROPERTY_TYPE_ID.name} = ANY(?) " +
                 "AND ${ENTITY_SET_ID.name} = ANY(?) " +
                 "AND length(${HASH.name}) = 16 " +
-                "ON CONFLICT ($keyCols) DO UPDATE SET " +
```
Comment (geekbeast) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt:140]: It looks like DO UPDATE SET is missing.

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt [hunk 2]
```diff
@@ -23,127 +19,201 @@ class UpdateDateTimePropertyHash(private val toolbox: Toolbox) : Upgrade {
...
+                "SELECT $insertSelectCols FROM $TEMP_TABLE_NAME " +
+                onConflictClause(DATA.name)
+    }
+
+    private fun deleteSql(): String {
+        val cteName = "to_delete"
+
+        val colsToMatch = (DATA_TABLE_KEY_COLS - HASH.name).joinToString(" AND ") { "${DATA.name}.$it = $cteName.$it" }
+
+        return "WITH $cteName AS ( SELECT * FROM $TEMP_TABLE_NAME ) " +
```
Comment (geekbeast) [inline src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: Don't use a CTE for delete as this breaks pushdown.

**Review by geekbeast — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt (hunk 2 above)
Comment (geekbeast) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt]: ```DELETE FROM ${DATA.name} USING $TEMP_TABLE_NAME where $colsToMatch AND ANY( $TEMP_TABLE_NAME.${OLD_HASH_COL.name})``` should do it right?

File: src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt (hunk 1 above)
Comment (katherinebernstein) [reply src/main/kotlin/com/openlattice/mechanic/upgrades/UpdateDateTimePropertyHash.kt:140]: no, it's in onConflict

**Review by geekbeast — APPROVED** [review]

**Pull request merged**